src/main/java/org/usfirst/frc4048/common/swervev2/SwerveDrivetrain.java

```diff
@@ -1,170 +1,102 @@
-        frontLeftDrive.setInverted(true);
-        frontRightDrive.setInverted(false);
-        backRightDrive.setInverted(false);
-        backLeftDrive.setInverted(true);
-        navxGyro.setAngleAdjustment(gyroOffset);
+    public SwerveDrivetrain(SwerveIdConfig frontLeftConfig, SwerveIdConfig frontRightConfig, SwerveIdConfig backLeftConfig, SwerveIdConfig backRightConfig,
+                            KinematicsConversionConfig conversionConfig, SwervePidConfig pidConfig, Gyro gyro) 
+    {
+        this.gyro = gyro;
+        EncodedSwerveSparkMax encodedSwerveSparkMaxFL = new EncodedSwerveMotorBuilder(frontLeftConfig, conversionConfig).build();
+        EncodedSwerveSparkMax encodedSwerveSparkMaxFR = new EncodedSwerveMotorBuilder(frontRightConfig, conversionConfig).build();
+        EncodedSwerveSparkMax encodedSwerveSparkMaxBL = new EncodedSwerveMotorBuilder(backLeftConfig, conversionConfig).build();
+        EncodedSwerveSparkMax encodedSwerveSparkMaxBR = new EncodedSwerveMotorBuilder(backRightConfig, conversionConfig).build();
+
+        this.frontLeft = new GenericSwerveModule(encodedSwerveSparkMaxFL, pidConfig.getDrivePid(),pidConfig.getSteerPid(),pidConfig.getDriveGain(),pidConfig.getSteerGain(),pidConfig.getGoalConstraint());
+        this.frontRight = new GenericSwerveModule(encodedSwerveSparkMaxFR, pidConfig.getDrivePid(),pidConfig.getSteerPid(),pidConfig.getDriveGain(),pidConfig.getSteerGain(),pidConfig.getGoalConstraint());
+        this.backLeft = new GenericSwerveModule(encodedSwerveSparkMaxBL, pidConfig.getDrivePid(),pidConfig.getSteerPid(),pidConfig.getDriveGain(),pidConfig.getSteerGain(),pidConfig.getGoalConstraint());
+        this.backRight = new GenericSwerveModule(encodedSwerveSparkMaxBR, pidConfig.getDrivePid(),pidConfig.getSteerPid(),pidConfig.getDriveGain(),pidConfig.getSteerGain(),pidConfig.getGoalConstraint());
+        this.frontRight.getSwerveMotor().getDriveMotor().setInverted(true);
+        this.frontLeft.getSwerveMotor().getDriveMotor().setInverted(false);
+        this.backRight.getSwerveMotor().getDriveMotor().setInverted(true);
+        this.backLeft.getSwerveMotor().getDriveMotor().setInverted(false);
     }
 
-    @Override
-    public void periodic() {
-        navxGyroValue = getGyro();
-    }
-    public float getRoll() {
-        return navxGyro.getRoll();
-    }
-    
-    private static final double MODULE_MAX_ANGULAR_ACCELERATION =
-            2 * Math.PI; // radians per second squared
+
     public void drive(double xSpeed, double ySpeed, double rot, boolean fieldRelative) {
         SwerveModuleState[] swerveModuleStates = kinematics.toSwerveModuleStates(
                 fieldRelative
-                        ? ChassisSpeeds.fromFieldRelativeSpeeds(xSpeed, ySpeed, rot, new Rotation2d(Math.toRadians(navxGyroValue)))
+                        ? ChassisSpeeds.fromFieldRelativeSpeeds(xSpeed, ySpeed, rot, new Rotation2d(Math.toRadians(getGyro())))
```

Comment:
IIUC this may slow the system down as the value was obtainer in the periodic before.

Reply:
done